src/mpl/src/gavl/mpl_gavl.c

```diff
@@ -182,6 +189,94 @@ int MPL_gavl_tree_create(void (*free_fn) (void *), MPL_gavl_tree_t * gavl_tree)
     goto fn_exit;
 }
 
+static void gavl_tree_insert_search(gavl_tree_s * gavl_tree_iptr, gavl_tree_node_s * node_ptr,
+                                    gavl_tree_node_s * cur_node, int *direction,
+                                    gavl_tree_node_s ** cur_node_ret,
+                                    gavl_tree_stack_s * node_stack)
+{
+    int cmp_ret;
+
+    do {
+        cmp_ret = gavl_subset_cmp_func((uintptr_t) node_ptr->addr, node_ptr->len, cur_node);
+        if (cmp_ret == SEARCH_LEFT) {
+            if (cur_node->left != NULL) {
+                STACK_PUSH(node_stack, cur_node);
+                cur_node = cur_node->left;
+                continue;
+            } else {
+                *direction = SEARCH_LEFT;
+                break;
+            }
+        } else if (cmp_ret == SEARCH_RIGHT) {
+            if (cur_node->right != NULL) {
+                STACK_PUSH(node_stack, cur_node);
+                cur_node = cur_node->right;
+                continue;
+            } else {
+                *direction = SEARCH_RIGHT;
+                break;
+            }
+        } else {
+            /* we cannot insert the duplicate buffer */
+            gavl_tree_iptr->gavl_free_fn((void *) node_ptr->val);
+            MPL_free(node_ptr);
+            *direction = NO_BUFFER_MATCH;
+            break;
+        }
+    } while (1);
+
+    *cur_node_ret = cur_node;
+    return;
+}
+
+static void gavl_tree_insert_node(int direction, gavl_tree_node_s * node_ptr,
+                                  gavl_tree_node_s * cur_node)
+{
+    if (direction == SEARCH_LEFT)
+        cur_node->left = node_ptr;
+    else
+        cur_node->right = node_ptr;
+    node_ptr->parent = cur_node;
+    return;
```

Comment:
This is too trivial to be a function. I would imagine the caller like this (psedocode):
```
    (parent, dir) = tree_search(tree, node->addr, node->len);
     if (dir == LEFT) {
         parent->left = node;
    } else if (dir == RIGHT) {
        parent->right = node;
    } else {
        free(parent);
        /* where do we insert node? */
    }
```
Search and then Insert and Free are at the same level of context, it is easier to read than tracing down the trivial functions.

Reply:
yeah, I just want to make it more modulized, and making them in a function would make `MPL_gavl_tree_insert` cleaner.
We don't free `parent` here. If we find no matched node, we will free `node` here. 
I can move it back.